Add product collections

brandon-deploy opened the pull request:
Fixes #4 

## Changes I have made:
- Added product collections that can be requested through the API

I couldn't get the forms to conditionally hide. Though, my best attempt is still there within the UI component if anyone want to give it a shot.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages

Commit:
Add product collections

mugureltanase commented on src/BuilderIOProductCollections/Block/Adminhtml/ProductCollection/Edit/Tab/Conditions.php:
```diff
@@ -0,0 +1,197 @@
+        $conditions->setJsFormObject($formName);
+    }
+
+    public function getTabLabel()
```
any reason why this function are empty ?

brandon-deploy replied:
It's required for the interface but has no functional use for it's implementation

mugureltanase commented on Helper/Data.php:
```diff
@@ -0,0 +1,204 @@
+
+        $collection = $this->productCollectionFactory->create()
+            ->addAttributeToSelect('entity_id')
+            ->addAttributeToFilter('status', ['in' => $this->scopeConfig->getValue('catalog/frontend/flat_catalog_product_status') ? [1] : [\Magento\Catalog\Model\Product\Status::STATUS_ENABLED]])
```
why not using a constant var instead of adding the full string ?

brandon-deploy replied:
because it isn't defined within this module

mugureltanase commented on Helper/Data.php:
```diff
@@ -0,0 +1,204 @@
+
+        $collection = $this->productCollectionFactory->create()
+            ->addAttributeToSelect(['name', 'price', 'small_image', 'url_key', 'sku'])
+            ->addAttributeToFilter('status', ['in' => [\Magento\Catalog\Model\Product\Status::STATUS_ENABLED]])
```
why not using the qualifier instead of import ? 

\Magento\Catalog\Model\Product\Status  ->

use Magento\Catalog\Model\Product\Status;  
Status::STATUS_ENABLE

brandon-deploy replied:
I will clear up fqns

mugureltanase commented on src/BuilderIOProductCollections/Model/ProductCollection.php:
```diff
@@ -0,0 +1,273 @@
+
+    protected int $productCount;
+
+    public function __construct(
```
this has strange format.

brandon-deploy replied:
Please be more specific

Commit:
add changes

Commit:
optimise imports and add url generation

nathanchick committed:
refactored to push more attributes over, added visibility restrictions and full image urls

nathanchick committed:
changed menu order

nathanchick committed:
fixed bug with adding new and saving

nathanchick committed:
code styles

nathanchick committed:
removed file

nathanchick committed:
tidied up some old code

nathanchick committed:
code styling

nathanchick committed:
unneeded

nathanchick commented:
Hey @brandon-deploy  👋 a couple of things I spotted:

1. The category checkboxes in the conditions aren’t working — they’re throwing a JS error.
2. The save button is redirecting back to the edit page. I traced this to the form-glue.js file, which seems to be adding the back param on submission. Nothing obvious jumped out, so might need a closer look.
3. The visibility based on the “type” selection isn’t working. Again, nothing obvious in the code, but needs some digging.
4. Removed Helper Class, probably the right thing to do, but there were many dependencies. I've quickly removed those so I could test the updates a little more. Please double-check that all is ok.

nathanchick reviewed (changes requested):
^^^

Commit:
fix chooser for conditions within product collections

Commit:
fix redirect

Commit:
fix visibility

Commit:
add back the product count saving

Commit:
remove uneeded collection method calls

Commit:
add visibility back

brandon-deploy commented:
Hi @nathanchick I have fixed the issues. The chooser (category id's form component thing) was expecting to be in a fieldset called rule_condition_fieldset due it being the same one from the catalog_rule module.

Commit:
split

Commit:
remove all none product collection related files and content from product collection

Commit:
add master

Commit:
declare strict everywhere

Commit:
Code inspection and fix namespace

Commit:
add modules to psr

I haven't tested this yet

Commit:
fix autoload

Commit:
bump version

Commit:
fix ProductCollection namespacing

Commit:
fix indexing

brandon-deploy commented:
Hi @nathanchick, I've split the modules as requested. I've also fixed a few things.

As far as functionality goes, I think this is ready to go.

let me know if you spot anything your side.

Commit:
fix

Commit:
linting

Commit:
fix 8.4 issues

Commit:
remove nested compsoer.json

Commit:
add 8.5 to compatibility
